--- /home/hamad/dev/linux/kernel/Kconfig.hz	2021-04-29 23:55:01.480064966 +0300
+++ kernel/Kconfig.hz	2021-05-04 15:46:47.387752520 +0300
@@ -46,6 +46,9 @@
 	 1000 Hz is the preferred choice for desktop systems and other
 	 systems requiring fast interactive responses to events.

+	config HZ_2000
+		bool "2000 HZ"
+
 endchoice

 config HZ
@@ -54,6 +57,7 @@
 	default 250 if HZ_250
 	default 300 if HZ_300
 	default 1000 if HZ_1000
+	default 2000 if HZ_2000

 config SCHED_HRTICK
 	def_bool HIGH_RES_TIMERS
