diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index 794c2cb945f8..b157314b088f 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -6712,18 +6712,18 @@ select_task_rq_fair(struct task_struct *p, int prev_cpu, int wake_flags)
 	/* SD_flags and WF_flags share the first nibble */
 	int sd_flag = wake_flags & 0xF;

-	if (wake_flags & WF_TTWU) {
-		record_wakee(p);
-
-		if (sched_energy_enabled()) {
-			new_cpu = find_energy_efficient_cpu(p, prev_cpu);
-			if (new_cpu >= 0)
-				return new_cpu;
-			new_cpu = prev_cpu;
-		}
-
-		want_affine = !wake_wide(p) && cpumask_test_cpu(cpu, p->cpus_ptr);
-	}
+	//if (wake_flags & WF_TTWU) {
+		//record_wakee(p);
+
+		//if (sched_energy_enabled()) {
+			//new_cpu = find_energy_efficient_cpu(p, prev_cpu);
+			//if (new_cpu >= 0)
+				//return new_cpu;
+			//new_cpu = prev_cpu;
+		//}
+
+		//want_affine = !wake_wide(p) && cpumask_test_cpu(cpu, p->cpus_ptr);
+	//}

 	rcu_read_lock();
 	for_each_domain(cpu, tmp) {
